File: apps/mobile/ios/App/Shared/WidgetStore.swift
import Foundation

/**
 * The two files a home-screen widget lives on, iOS side (plan Widgets, W2).
 *
 * The counterpart of `WidgetStore.java`, and the same two rules. `snapshot.json`
 * is what the app worked out the last time it ran — no JavaScript runs while
 * Plainva is closed, so a widget can compute nothing and only draws this.
 * `actions.json` runs the other way: a tick made on the home screen is an
 * ORDER, which the app redeems when it next opens, because ticking a task off
 * means the database's completion model, the recurrence and the provider sync,
 * and none of that exists in Swift.
 *
 * Unlike Android, the widget here is a SEPARATE PROCESS. Both files therefore
 * live in the App Group container, beside `share-inbox-v1`, and every access
 * takes the same `flock` the share queue uses — the extension and the app can
 * genuinely write at the same moment.
 *
 * Every public method takes that lock exactly ONCE, around private helpers
 * that never take it. `flock` is not reentrant the way a Java monitor is: a
 * second `open` + `LOCK_EX` from the same process waits for a lock the same
 * thread is holding, and the widget hangs rather than drawing. So the rule
 * here is structural — a `_`-prefixed helper assumes the lock is held and
 * never calls a public method.
 *
 * File protection is `completeUntilFirstUserAuthentication`, not `complete`:
 * a lock-screen widget has to render while the device is locked. What that
 * admits is bounded by what the snapshot carries — a title, a day, a time, an
 * index; never note text, never a path, and nothing at all while the
 * workspace is sealed.
 */
public enum WidgetStoreFailure: Error {
    case storage
}

public final class WidgetStore {
    /** A tick older than this was made against a snapshot nobody can resolve any more. */
    static let actionTTLms: Int64 = 7 * 24 * 60 * 60 * 1000
    /** A bound on a file two processes read: orders pile up only while the app is never opened. */
    static let maxActions = 200
    static let maxSnapshotBytes = 256 * 1024

    public static let appGroup = "group.com.plainva.app"

    let root: URL
    private let fm = FileManager.default

    public init(root: URL? = nil) throws {
        guard let location = root ?? FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: WidgetStore.appGroup)?.appendingPathComponent("widget-v1", isDirectory: true) else {
            throw WidgetStoreFailure.storage
        }
        self.root = location
        try fm.createDirectory(at: location, withIntermediateDirectories: true)
        #if os(iOS)
        try? fm.setAttributes([.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication], ofItemAtPath: location.path)
        #endif
        var url = location
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        try? url.setResourceValues(values)
    }

    private var snapshotURL: URL { root.appendingPathComponent("snapshot.json") }
    private var actionsURL: URL { root.appendingPathComponent("actions.json") }

    /** One lock file for both, so a tick and a fresh snapshot never interleave. */
    private func locked<T>(_ action: () throws -> T) throws -> T {
        let fd = open(root.appendingPathComponent(".lock").path, O_CREAT | O_RDWR, 0o600)
        guard fd >= 0 else { throw WidgetStoreFailure.storage }
        defer { close(fd) }
        guard flock(fd, LOCK_EX) == 0 else { throw WidgetStoreFailure.storage }
        defer { flock(fd, LOCK_UN) }
        return try action()
    }

    // ------------------------------------------------- helpers, lock held

    /** Temp file, fsync, rename: a kill mid-write never leaves a half-file. */
    private func _write(_ url: URL, _ text: String) throws {
        let temporary = root.appendingPathComponent("." + url.lastPathComponent + "-" + UUID().uuidString)
        defer { if fm.fileExists(atPath: temporary.path) { try? fm.removeItem(at: temporary) } }
        guard let data = text.data(using: .utf8) else { throw WidgetStoreFailure.storage }
        #if os(iOS)
        try data.write(to: temporary, options: [.completeFileProtectionUntilFirstUserAuthentication])
        #else
        try data.write(to: temporary)
        #endif
        let handle = try FileHandle(forWritingTo: temporary)
        defer { try? handle.close() }
        if fcntl(handle.fileDescriptor, F_FULLFSYNC) != 0 {
            guard fsync(handle.fileDescriptor) == 0 else { throw WidgetStoreFailure.storage }
        }
        guard rename(temporary.path, url.path) == 0 else { throw WidgetStoreFailure.storage }
    }

    private func _read(_ url: URL) -> String? {
        guard let size = try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize, size <= WidgetStore.maxSnapshotBytes else { return nil }
        return try? String(contentsOf: url, encoding: .utf8)
    }

    private func _object(_ url: URL) -> [String: Any]? {
        guard let raw = _read(url), let data = raw.data(using: .utf8) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }

    private func _writtenAt() -> Int64 {
        ((_object(snapshotURL)?["writtenAt"]) as? NSNumber)?.int64Value ?? 0
    }

    private func _actions() -> [[String: Any]] {
        guard let root = _object(actionsURL), root["version"] as? Int == 1,
              let actions = root["actions"] as? [[String: Any]] else { return [] }
        return actions
    }

    private func _nextId() -> Int64 {
        max(1, ((_object(actionsURL)?["nextId"]) as? NSNumber)?.int64Value ?? 1)
    }

    private func _saveActions(_ actions: [[String: Any]], nextId: Int64) throws {
        let root: [String: Any] = ["version": 1, "nextId": NSNumber(value: nextId), "actions": actions]
        let data = try JSONSerialization.data(withJSONObject: root, options: [.sortedKeys])
        guard let text = String(data: data, encoding: .utf8) else { throw WidgetStoreFailure.storage }
        try _write(actionsURL, text)
    }

    // ------------------------------------------------------------- snapshot

    public func writeSnapshot(_ json: String) throws {
        try locked { try _write(snapshotURL, json) }
    }

    public func readSnapshot() -> String? {
        (try? locked { _read(snapshotURL) }) ?? nil
    }

    public func clearSnapshot() {
        try? locked { try? fm.removeItem(at: snapshotURL) }
    }

    /** When the app wrote the snapshot on disk — what an order has to name. */
    public func snapshotWrittenAt() -> Int64 {
        (try? locked { _writtenAt() }) ?? 0
    }

    // --------------------------------------------------------------- orders

    /**
     * Records a tick against the snapshot currently on disk; answers the
     * order's id, or 0 when there was nothing to record.
     */
    @discardableResult
    public func enqueue(index: Int, at: Int64) -> Int64 {
        (try? locked { () -> Int64 in
            let snapshotAt = _writtenAt()
            guard snapshotAt != 0, index >= 0 else { return 0 }
            var kept: [[String: Any]] = []
            for action in _actions() {
                // The same row tapped twice is one tick, not two.
                if (action["snapshotAt"] as? NSNumber)?.int64Value == snapshotAt, (action["index"] as? NSNumber)?.intValue == index {
                    return (action["id"] as? NSNumber)?.int64Value ?? 0
                }
                // An order whose snapshot is long gone can no longer name a row.
                let made = (action["at"] as? NSNumber)?.int64Value ?? 0
                if at - made > WidgetStore.actionTTLms { continue }
                kept.append(action)
            }
            guard kept.count < WidgetStore.maxActions else { return 0 }
            let id = _nextId()
            kept.append(["id": NSNumber(value: id), "index": NSNumber(value: index), "snapshotAt": NSNumber(value: snapshotAt), "at": NSNumber(value: at)])
            try _saveActions(kept, nextId: id + 1)
            return id
        }) ?? 0
    }

    /** Every order still waiting, as the bridge hands it to JavaScript. */
    public func pendingActions() -> [[String: Any]] {
        (try? locked { _actions() }) ?? []
    }

    /** Drops the orders the app has redeemed — by id, never wholesale. */
    public func clearActions(ids: Set<Int64>) {
        try? locked {
            // A tap between the app's read and this call has to survive it.
            let kept = _actions().filter { !ids.contains(($0["id"] as? NSNumber)?.int64Value ?? -1) }
            try? _saveActions(kept, nextId: _nextId())
        }
    }

    /**
     * Which rows of the CURRENT snapshot carry a tick the app has not redeemed
     * — what lets the widget show the tick the moment it is made.
     */
    public func pendingIndexes() -> Set<Int> {
        (try? locked { () -> Set<Int> in
            let snapshotAt = _writtenAt()
            guard snapshotAt != 0 else { return [] }
            var out: Set<Int> = []
            for action in _actions() where (action["snapshotAt"] as? NSNumber)?.int64Value == snapshotAt {
                if let index = (action["index"] as? NSNumber)?.intValue, index >= 0 { out.insert(index) }
            }
            return out
        }) ?? []
    }
}

File: apps/mobile/android/app/src/main/java/com/plainva/app/WidgetStore.java
package com.plainva.app;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.HashSet;
import java.util.Set;

/**
 * The two files a home-screen widget lives on (plan Widgets, W2).
 *
 * No JavaScript runs while Plainva is closed, so a widget can work nothing
 * out: it draws what the app wrote down the last time it ran. That file is
 * {@code snapshot.json}.
 *
 * The other direction is {@code actions.json}. Ticking a task off means
 * setting a status by the database's completion model, carrying a recurrence
 * forward and syncing the change to a provider — none of which exists in
 * Java, and rebuilding it here would be a second truth. So a tick on the home
 * screen is only an ORDER: the widget shows the tick at once, the app redeems
 * it the next time it opens.
 *
 * Both files live in the app's private storage. On Android that is enough for
 * a widget to read them, because an {@code AppWidgetProvider} runs in the
 * app's own process — unlike iOS, where the widget is a separate process and
 * the same two files have to live in the App Group container.
 *
 * Every write goes through a temp file and a rename, so a process killed
 * mid-write never leaves a half-file that the widget would then fail to parse
 * — and a widget that cannot parse its file has nothing else to show.
 */
public final class WidgetStore {
    private WidgetStore() {}

    private static final String DIR = "widget";
    private static final String SNAPSHOT = "snapshot.json";
    private static final String ACTIONS = "actions.json";

    /** A tick older than this was made against a snapshot nobody can resolve any more. */
    private static final long ACTION_TTL_MS = 7L * 24 * 60 * 60 * 1000;
    /** A bound on a file two processes read: taps accumulate only while the app is never opened. */
    private static final int MAX_ACTIONS = 200;
    /** The snapshot itself is capped at 60 rows by the writer; this is the file-level backstop. */
    private static final int MAX_SNAPSHOT_BYTES = 256 * 1024;

    /** One lock for both files: the provider and the bridge are in the same process. */
    private static final Object LOCK = new Object();

    private static File directory(Context context) {
        File dir = new File(context.getFilesDir(), DIR);
        if (!dir.exists()) dir.mkdirs();
        return dir;
    }

    // ---------------------------------------------------------------- files

    /**
     * Temp file plus rename. {@code File.renameTo} is {@code rename(2)} on the
     * same filesystem, which is atomic — and available on every API level this
     * app supports, unlike {@code java.nio.file.Files}.
     */
    private static void writeAtomically(File target, String content) {
        File dir = target.getParentFile();
        File temp = new File(dir, "." + target.getName() + ".tmp");
        try {
            FileOutputStream out = new FileOutputStream(temp);
            try {
                out.write(content.getBytes(StandardCharsets.UTF_8));
                out.flush();
                out.getFD().sync();
            } finally {
                out.close();
            }
            if (!temp.renameTo(target)) {
                // A failed rename must not leave the stale temp file behind.
                temp.delete();
            }
        } catch (Exception e) {
            temp.delete();
        }
    }

    private static String readFile(File file) {
        if (!file.isFile() || file.length() > MAX_SNAPSHOT_BYTES) return null;
        try {
            FileInputStream in = new FileInputStream(file);
            try {
                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
                byte[] chunk = new byte[8192];
                int read;
                while ((read = in.read(chunk)) > 0) buffer.write(chunk, 0, read);
                return new String(buffer.toByteArray(), StandardCharsets.UTF_8);
            } finally {
                in.close();
            }
        } catch (Exception e) {
            return null;
        }
    }

    // ------------------------------------------------------------- snapshot

    /** What the app last worked out, as the bytes the widget draws. */
    public static void writeSnapshot(Context context, String json) {
        synchronized (LOCK) {
            writeAtomically(new File(directory(context), SNAPSHOT), json);
        }
    }

    /** The snapshot as written, or null when there is none yet or it is unreadable. */
    public static String readSnapshot(Context context) {
        synchronized (LOCK) {
            return readFile(new File(directory(context), SNAPSHOT));
        }
    }

    /** Wipes the snapshot — used when the vault is locked, changed or removed (W6). */
    public static void clearSnapshot(Context context) {
        synchronized (LOCK) {
            new File(directory(context), SNAPSHOT).delete();
        }
    }

    /**
     * When the app wrote the current snapshot. An index only means a row
     * within ONE snapshot, so this is what an order has to name.
     */
    public static long snapshotWrittenAt(Context context) {
        String raw = readSnapshot(context);
        if (raw == null) return 0L;
        try {
            return new JSONObject(raw).optLong("writtenAt", 0L);
        } catch (Exception e) {
            return 0L;
        }
    }

    // --------------------------------------------------------------- orders

    private static JSONArray readActions(Context context) {
        String raw = readFile(new File(directory(context), ACTIONS));
        if (raw == null) return new JSONArray();
        try {
            JSONObject root = new JSONObject(raw);
            if (root.optInt("version", 0) != 1) return new JSONArray();
            JSONArray actions = root.optJSONArray("actions");
            return actions == null ? new JSONArray() : actions;
        } catch (Exception e) {
            return new JSONArray();
        }
    }

    private static void saveActions(Context context, JSONArray actions, long nextId) {
        try {
            JSONObject root = new JSONObject();
            root.put("version", 1);
            root.put("nextId", nextId);
            root.put("actions", actions);
            writeAtomically(new File(directory(context), ACTIONS), root.toString());
        } catch (Exception e) {
            /* an order that cannot be written is one the app never sees — better than a crash on a home screen */
        }
    }

    private static long nextId(Context context) {
        String raw = readFile(new File(directory(context), ACTIONS));
        if (raw == null) return 1L;
        try {
            return Math.max(1L, new JSONObject(raw).optLong("nextId", 1L));
        } catch (Exception e) {
            return 1L;
        }
    }

    /**
     * Records a tick made on the home screen against the snapshot currently on
     * disk. Returns the order's id, or 0 when there was nothing to record.
     */
    public static long enqueue(Context context, int index, long at) {
        synchronized (LOCK) {
            long snapshotAt = snapshotWrittenAt(context);
            if (snapshotAt == 0L || index < 0) return 0L;
            JSONArray existing = readActions(context);
            JSONArray kept = new JSONArray();
            for (int i = 0; i < existing.length(); i++) {
                JSONObject action = existing.optJSONObject(i);
                if (action == null) continue;
                // The same row tapped twice is one tick, and an order whose
                // snapshot is long gone can no longer name a row.
                if (action.optLong("snapshotAt") == snapshotAt && action.optInt("index", -1) == index) return action.optLong("id");
                if (at - action.optLong("at") > ACTION_TTL_MS) continue;
                kept.put(action);
            }
            if (kept.length() >= MAX_ACTIONS) return 0L;
            long id = nextId(context);
            try {
                JSONObject action = new JSONObject();
                action.put("id", id);
                action.put("index", index);
                action.put("snapshotAt", snapshotAt);
                action.put("at", at);
                kept.put(action);
            } catch (Exception e) {
                return 0L;
            }
            saveActions(context, kept, id + 1);
            return id;
        }
    }

    /** Every order still waiting, as the bridge hands it to JavaScript. */
    public static JSONArray pendingActions(Context context) {
        synchronized (LOCK) {
            return readActions(context);
        }
    }

    /** Drops the orders the app has redeemed. Clearing by id, never "all". */
    public static void clearActions(Context context, Set<Long> ids) {
        synchronized (LOCK) {
            JSONArray existing = readActions(context);
            JSONArray kept = new JSONArray();
            for (int i = 0; i < existing.length(); i++) {
                JSONObject action = existing.optJSONObject(i);
                // A tap that lands between the app READING the queue and
                // clearing it must survive, so only the named ids go.
                if (action != null && !ids.contains(action.optLong("id"))) kept.put(action);
            }
            saveActions(context, kept, nextId(context));
        }
    }

    /**
     * Which rows of the CURRENT snapshot carry a tick that the app has not
     * redeemed yet — what lets the widget show the tick the moment it is made.
     */
    public static Set<Integer> pendingIndexes(Context context) {
        Set<Integer> out = new HashSet<Integer>();
        synchronized (LOCK) {
            long snapshotAt = snapshotWrittenAt(context);
            if (snapshotAt == 0L) return out;
            JSONArray actions = readActions(context);
            for (int i = 0; i < actions.length(); i++) {
                JSONObject action = actions.optJSONObject(i);
                if (action == null || action.optLong("snapshotAt") != snapshotAt) continue;
                int index = action.optInt("index", -1);
                if (index >= 0) out.add(Integer.valueOf(index));
            }
        }
        return out;
    }

    // --------------------------------------------------------------- redraw

    /**
     * Asks every widget of this package to redraw.
     *
     * Two steps, because an update broadcast redraws a widget's own views but
     * does NOT tell a list adapter that its data changed — the rows would keep
     * showing the previous snapshot until the launcher happened to rebind.
     * The package-scoped broadcast reaches whatever providers exist; the
     * explicit call reaches the one that has a list.
     */
    public static void reload(Context context) {
        try {
            Intent intent = new Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
            intent.setPackage(context.getPackageName());
            context.sendBroadcast(intent);
        } catch (Exception e) {
            /* a redraw that does not happen costs a stale widget, never a crash */
        }
        try {
            TodayWidgetProvider.refresh(context);
        } catch (Exception e) {
            /* same */
        }
    }
}
